sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -1292,10 +1323,24 @@ class Dataset[T] private[sql](
         colRegex(colName)
       } else {
         val expr = resolve(colName)
-        Column(expr)
+        Column(addDataFrameIdToCol(expr))
```

Comment:
As indicated by the change in `Analyzer`, after this two attributes we considered the same, now are different if we compare it without canonicalization.

Not sure if it will be an issue.

Reply:
Internally, I think we should never compare 2 attributes by `equal` not `semanticEqual`.

Externally, if users rely on the equality of `Column`, yes this PR may break it under some cases. If this matters, I can add a migration guide or update `Column.equal`.

Reply:
I feel that metadata is not well aware. It might be confusing to some people that the comparison fails if they know it works before. An update to migration guide or `Column.equal` both sounds good.